### DISCUSSION TITLE
Allow actor passed to useSelector to be undefined

### DISCUSSION THREAD

**Discussion by SimeonC** [statelyai/xstate#3159] [Ideas] [answered]
I'd like to discuss about whether this is something the main package would be interested updated to support (or I just copy useSelector logic and modify it for my use only).

```diff
export function useSelector<
  TActor extends ActorRef<any, any>,
  T,
  TEmitted = TActor extends Subscribable<infer Emitted> ? Emitted : never
>(
-  actor: TActor,
-  selector: (emitted: TEmitted) => T,
+  actor: TActor,
+  selector: (emitted: TEmitted | undefined) => T,
  compare: (a: T, b: T) => boolean = defaultCompare,
-  getSnapshot: (a: TActor) => TEmitted = defaultGetSnapshot
+  getSnapshot: (a: TActor | undefined) => TEmitted = defaultGetSnapshot
) {
```

https://github.com/statelyai/xstate/blob/b36ef9dda560fca4c00428f48742fd9d2e325324/packages/xstate-react/src/useSelector.ts#L20-L29

## Example Usage

This solution I came up with to try to work around the deprecation of TypeStates and "Lazy" options. I used to do this by utilising a typestate like;

```ts
type TypeState = {
  value: 'loading';
  context: Partial<Context>;
} | {
  value: 'running';
  context: Context;
}
```

```tsx
const lazyOptions: (() => Options | Promise<Options>) | Options | Promise<Options>;
const preloadService = useInterpret(
  () =>
    createMachine(
      {
        schema: {
          context: {} as {
            service?: InterpreterFrom<...>; // in this example I'm intentionally not spawning as the sub-service never needs to talk to this one
          },
          services: {} as {
            'Resolve Options': {
              data: Options;
            };
          }
        },
        initial: 'Load',
        id: 'Load Lazy Options',
        states: {
          Load: {
            invoke: {
              src: 'Resolve Options',
              onDone: [
                {
                  actions: 'Invoke Actor',
                  target: '#Load Lazy Options.Rest'
                }
              ],
              onError: [
                {
                  target: '#Load Lazy Options.Error'
                }
              ]
            }
          },
          Rest: {},
          Error: {}
        }
      },
      {
        services: {
          'Resolve Options': () => Promise.resolve(
                typeof lazyOptions === 'function'
                  ? lazyOptions()
                  : lazyOptions
              )
        },
        actions: {
          'Invoke Tab Form Actor': assign({
            tabFormService: (_, event) =>
              interpret(
                someMachine(event.data),
                { devTools: CONFIG.useDevTools }
              ).start(),
        }
      }
    ),
  { devTools: CONFIG.useDevTools }
);
const { service } = useSelector(
  preloadService,
  (state) => state.context
);
const someValue = useSelector(service, (state) => state?.context.someValue);
```

**Comment by Andarist** [comment, marked as answer]
I feel like a good chunk of my other post from today goes into my thoughts on "accepting actor argument to be undefined". You can find it [here](https://github.com/statelyai/xstate/issues/3156#issuecomment-1070829935). I would appreciate your thoughts on this - let's keep them in this discussion here though.

I'm also unsure what exactly you want to solve here and why. You have mentioned something like this:
> This solution I came up with to try to work around the deprecation of TypeStates and "Lazy" options.

But this is somewhat enigmatic to me. Could you showcase the code that you had before and compare it to what you have to do now to achieve the same thing~?

**Reply by SimeonC** [reply]
Sure, I think a lot of my recent typescript/xstate problems actually come from trying to use machines inside custom hooks rather than directly. I have started moving to "inside types" and "outside types" which does kinda make things easier. 

This is my pre 4.30 version (removed some logic that's not directly necessary). I was never quite happy about the "Pretend this partial context I'm passing you is a real context" approach.

My biggest pain point is this snippet from the return (I could work around the rest via doubly nested components).

```tsx
useIsDirty: () => {
        const isFormDirty = useTabFormSelector(
          tabFormService,
          (state) =>
            // non-form states are pre-loading or final end states which are "pristine" in essence
            // as they can no longer be interacted with
            !!state &&
            state.matches('form') &&
            !state.matches({ form: 'pristine' })
        );
        return (
          tabFormService?.status !== InterpreterStatus.Stopped && isFormDirty
        );
      }
```

```tsx
export function useTabForm<
  TFieldValues extends FieldValues,
  TSubmitReturn extends any,
  TFormStructure extends FormStructure<
    TFieldValues,
    TFunctionArguments
  > = ExtendableFormStructure<TFieldValues>,
  TTabType extends TFunctionArguments = TFormStructure extends FormStructure<
    TFieldValues,
    infer ITabType
  >
    ? ITabType
    : never
>(
  // if these params are separate the type inference works correctly. Doesn't work when this is part of the options object
  initialFormData: TFieldValues | undefined,
  submitForm: OnSubmit<TFieldValues, TSubmitReturn>,
  options: Omit<
    Options<TFieldValues, TSubmitReturn, TTabType>,
    'initialFormData' | 'submitForm'
  > &
    HookOptions<TFieldValues>
): {
  renderFreeFormField: (name: FieldPath<TFieldValues>) => React.ReactNode;
  renderTabsAndContent: () => React.ReactNode;
  renderSubmitButton: () => React.ReactNode;
  useIsDirty: () => boolean;
  formService: Service<TFieldValues, TSubmitReturn, TFormStructure, TTabType>;
} {
  const service = useInterpret(() =>
    createTabFormMachine<TFieldValues, TSubmitReturn, TFormStructure, TTabType>(
      {
        ...options,
        initialFormData,
        submitForm
      }
    )
  );
  const freeFormMap = React.useMemo(
    () => ...);
  return React.useMemo(
    () => ({
      renderFreeFormField: (name: FieldPath<TFieldValues>) =>
        renderField(service, freeFormMap[name]),
      renderTabsAndContent: () => (
        <TabsAndContent
          service={service}
          getTranslatableLocales={options.getTranslatableLocales}
        />
      ),
      renderSubmitButton: () => (
        <SaveButton<TFieldValues, TSubmitReturn, TFormStructure, TTabType>
          service={service}
        />
      ),
      useIsDirty: () => {
        const isFormDirty = useSelector(
          service,
          (state) => !state.matches({ form: 'pristine' })
        );
        return service.status !== InterpreterStatus.Stopped && isFormDirty;
      },
      formService: service
    }),
    [service, freeFormMap, options.getTranslatableLocales]
  );
}

export function createTabFormMachine<
  TFieldValues extends FieldValues,
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  TSubmitReturn extends any,
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  TFormStructure extends FormStructure<TFieldValues, any>,
  TTabType extends string = TFormStructure extends FormStructure<
    TFieldValues,
    infer ITabType
  >
    ? ITabType
    : never
>(
  options: Options<TFieldValues, TSubmitReturn, TTabType>
): TabFormMachine<TFieldValues, TSubmitReturn, TFormStructure, TTabType> {
  const {
    structure,
    submitMessage,
    submitErrorMessage,
    getFieldErrorsPrompt,
    onComplete
  } = options;
  return createMachine(
    {
      schema: {
        context: {} as Context<TFieldValues, TTabType>,
        events: {} as Event<TFieldValues, TSubmitReturn, TTabType>
      },
      initial: 'preload',
      context: {
        structure,
        fieldToggles: {},
        displayValues: {}
      } as Context<TFieldValues, TTabType>,
      // this is a small shortcut - due to how preload is written context will always be correct in usage
      on: {
        SET_DISPLAY_VALUE: { actions: ['setDisplayValue'] }
      },
      states: {
        preload: {
          invoke: {
            id: 'loadInitialFormData',
            src: 'loadInitialFormData',
            onDone: {
              target: 'form',
              actions: ['initContext']
            },
            onError: {
              target: 'loadError',
              actions: [
                (ctx, event) => console.error('load error', { ctx, event })
              ]
            }
          }
        },
        loadError: {
          type: 'final'
        },
        form: {
          id: 'form',
          ...form stuff
```

My new attempt to get this working; (Main change is that the `createMachine` machine no longer has the "preload" state and odd case casting).

Note that `useTabFormSelector` is essentially the same as `useSelector` except I have "is actor defined" checks in place to prevent errors. (I copied the source of useSelector and scattered `?` around)
```tsx
export function useTabForm<
  TFieldValues extends FieldValues,
  TSubmitReturn,
  TFormStructure extends FormStructure<
    TFieldValues,
    string
  > = ExtendableFormStructure<TFieldValues>,
  TTabType extends string = TFormStructure extends FormStructure<
    TFieldValues,
    infer ITabType
  >
    ? ITabType
    : never
>(
  // if these params are separate the type inference works correctly. Doesn't work when this is part of the options object
  initialFormData: LazyValue<TFieldValues | undefined>,
  options: LazyValue<
    Omit<Options<TFieldValues, TSubmitReturn, TTabType>, 'initialFormData'> &
      HookOptions<TFieldValues>
  >
): {
  renderFreeFormField: (name: FieldPath<TFieldValues>) => React.ReactNode;
  renderTabsAndContent: () => React.ReactNode;
  renderSubmitButton: () => React.ReactNode;
  useIsDirty: () => boolean;
} {
  type TFreeFormMap = Record<
    FieldPath<TFieldValues>,
    NonNullable<
      Options<TFieldValues, TTabType, never>['freeFormFields']
    >[number]
  >;
  const preloadService = useInterpret(
    () =>
      /** @xstate-layout N4IgpgJg5mDOIC5QBkD2BDCACAKugRlgGKoBOAtlsugF4CeWA8gA4AuAlqgHawB0amAMQRuYXuy4A3VAGsxA7HkIkKVWgxYdufBQgnSAxui1cA2gAYAuolDNUsdiZsgAHogBMATgAsvbwFYAZgBGd3cADnCANgB2KKjvbwAaEDpET3NeT2zPcPMo8xjPGO9wgF8ylIVcAmIySmp6JjZOHn4MCEEwUlIyXmYAG2MAM3r2zBrlerUmzVadDr0pVCMTC2skEDsHJ023BC9fAJCwyNj4xJS0g-NMnOKvW8iYyIrKkC5UCDhnaqU61SNDQtbTjCDObaOVrOfbedxXRDBQq8YLxKKBTzuGIxfzmCLld5-WoqBrqZomPgAJTgrAh9ih3BhiCCwV4CXyIX8kRCUXhqURwX8WRy0QC5kSGO8FSqHUmANJsxBbQAoj0yHSdtC9syQmzvBzBdzUXzru48cLsgkop4ov5rf53NKQESpoCyXNtBqGVwmQhguEEX6YhabeFMYFzIFbaU3mUgA */
      createMachine(
        {
          tsTypes: {} as import('./index.typegen').Typegen0,
          schema: {
            context: {} as {
              getTranslatableLocales?: HookOptions<TFieldValues>['getTranslatableLocales'];
              freeFormMap?: TFreeFormMap;
              tabFormService?: InterpreterFrom<
                ReturnType<typeof createTabFormMachine>
              >;
              startPrompt?: StartPromptConfig;
            },
            services: {} as {
              'Resolve Options': {
                data: Options<TFieldValues, TSubmitReturn, TTabType> &
                  Omit<HookOptions<TFieldValues>, 'startPrompt'> & {
                    startPrompt?: StartPromptConfig;
                    context: ContextFrom<
                      TabFormMachine<TFieldValues, TFormStructure, TTabType>
                    >;
                  };
              };
            }
          },
          initial: 'Load',
          id: 'Load Tab Form Lazy Options',
          states: {
            Load: {
              invoke: {
                src: 'Resolve Options',
                onDone: [
                  {
                    actions: 'Invoke Tab Form Actor',
                    target: '#Load Tab Form Lazy Options.Rest'
                  }
                ],
                onError: [
                  {
                    target: '#Load Tab Form Lazy Options.Error'
                  }
                ]
              }
            },
            Rest: {},
            Error: {}
          }
        },
        {
          services: {
            'Resolve Options': async () => {
              try {
                const [resolvedInitialFormData, resolvedOptions] =
                  await Promise.all([
                    Promise.resolve(
                      typeof initialFormData === 'function'
                        ? // Typescript seems to have trouble figuring out this one
                          (initialFormData as Function)()
                        : initialFormData
                    ),
                    Promise.resolve(
                      typeof options === 'function' ? options() : options
                    )
                  ]);
                const { structure, freeFormFields, startPrompt } =
                  resolvedOptions;
                const finalInitialFormData = await composeInitialFormData({
                  initialFormData: resolvedInitialFormData,
                  structure,
                  freeFormFields
                });
                return {
                  ...resolvedOptions,
                  startPrompt: await Promise.resolve(
                    typeof startPrompt === 'function'
                      ? startPrompt(
                          getI18nextInstance().t,
                          finalInitialFormData
                        )
                      : startPrompt
                  ),
                  initialFormData: finalInitialFormData,
                  context: await buildContext<
                    TFieldValues,
                    TSubmitReturn,
                    TFormStructure,
                    TTabType
                  >({
                    structure,
                    initialFormData: finalInitialFormData,
                    freeFormFields
                  })
                };
              } catch (e) {
                console.error('Error Resolving Options', e, {
                  initialFormData,
                  options
                });
                throw e;
              }
            }
          },
          actions: {
            'Invoke Tab Form Actor': assign({
              tabFormService: (_, event) =>
                interpret(createTabFormMachine(event.data as never), {
                  devTools: CONFIG.useDevTools
                }).start(),
              getTranslatableLocales: (_, event) =>
                event.data.getTranslatableLocales,
              startPrompt: (_, event) => event.data.startPrompt,
              freeFormMap: (_, event) => ...
            })
          }
        }
      ),
    { devTools: CONFIG.useDevTools }
  );
  const { tabFormService, getTranslatableLocales, freeFormMap } = useSelector(
    preloadService,
    (state) => state.context
  );
  return React.useMemo(
    () => ({
      renderFreeFormField: (name: FieldPath<TFieldValues>) =>
        tabFormService && freeFormMap
          ? renderFormElement(tabFormService, freeFormMap[name])
          : null,
      renderTabsAndContent: () => (
        <TabsAndContent
          service={tabFormService}
          getTranslatableLocales={getTranslatableLocales}
        />
      ),
      renderSubmitButton: () => <SaveButton service={tabFormService} />,
      useIsDirty: () => {
        const isFormDirty = useTabFormSelector(
          tabFormService,
          (state) =>
            // non-form states are pre-loading or final end states which are "pristine" in essence
            // as they can no longer be interacted with
            !!state &&
            state.matches('form') &&
            !state.matches({ form: 'pristine' })
        );
        return (
          tabFormService?.status !== InterpreterStatus.Stopped && isFormDirty
        );
      }
    }),
    [tabFormService, getTranslatableLocales, freeFormMap]
  );
}
```

**Reply by SimeonC** [reply]
Eh, I did work around this in the end without the need for any hacks. I ended up doing a manual onTransition handler when I create the interpreter for the difficult `isDirty` variable (and removed the useMemo).

```tsx
tabFormService: (_, event) => {
                const service = interpret(
                  createTabFormMachine(event.data as never),
                  {
                    devTools: CONFIG.useDevTools
                  }
                );
                service.onStop(() => {
                  setIsDirty(false);
                });
                service.onTransition((state) => {
                  setIsDirty(
                    state.matches('form') &&
                      !state.matches({ form: 'pristine' })
                  );
                });
                return service.start();
              },
```

**Reply by Andarist** [reply]
Could this be described as wanting to provide custom hooks that encapsulate machines in a way that a consumer of such custom hook doesn't need to know about internals being a machine AND wanting to select from a child actor contained in that custom hook?

**Reply by SimeonC** [reply]
Yes, that is probably it in a nutshell. (I did have to read it about 3 times to make sure I understood you 😆 )

I like this approach of making custom hooks to encapsulate machines because it's a nice clean way of reusing code without all the bootstrapping. The downside is that with the new typegen I've found it quite hard to use Generics in the context/events (form data structure in this example). We have a 300 max-line eslint rule at work so I often have to split a machine into multiple file which means that `typeof machine` becomes un-generic.

I work around that by mangling the types at the hook border - outside the hook has good generic types, inside has static types. The disadvantage of this is that I haven't figured out how to pass the typegen out through the generics, though I'm tending to try avoid that situation and basically make the "Options" of the hook generic and properly typed and fudge some types on the output of the hook that don't need the generic inference.

**Reply by Andarist** [reply]
> Yes, that is probably it in a nutshell. (I did have to read it about 3 times to make sure I understood you 😆 )

Sorry for using a somewhat lengthy description for this. I've tried to be precise - to ensure that we are on the same page. Quite frankly, I've only skimmed through your code examples as they are quite lengthy and I wasn't completely sure what I should focus on there.

> I like this approach of making custom hooks to encapsulate machines because it's a nice clean way of reusing code without all the bootstrapping.

I think that a reasonable workaround for this would be for you to implement something like this in such a hook:
```js
const state = useSelector(actorRef || fallbackActorRef, state => state?.context.value)
```
Take a look at this [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOE7KMdGCmCAQWUagCU0AzCp4gTwCiANzTEYAeQBGAKzSs4AXzhcoEEHADkADwDOMLGk1wA9Cbgx+YNLoBccACwA6AMwAGJwFYAUKEixEOABXXTQAZTQAG3k2JRU1DU0AAT0DbBN0fBhjMwsrG3sAJicPAEZvb2QIYn04DQBeFEzsPEISNAAKBG84XpRq7G0Ye26+sf6g8Xs3Hr7Fb0UASgrcmRD4EAwAazQLIl04CDAYYGqMSODdEgBzOAwVSYE4AnwtuFsABUrq2qzoOEaeBgBCcUAwxAAJuoOos4AA+OAeTxwAD8lHEaBodA6IFh9kmEO47Qh3xq8C450iUleLFijUYtOgnC4AB4+EJROJpHJWBQCUTSBC4V1Zr1QpCYYhFGQFstvGhtP54FUyXB9IYAZdwlEYtAOn8oHAAD5GlSU6nILaMqAUdXYAEIu1oFFOFWDGCuiCTGDLXLjcYAPRRQA). Note that there might be some runtime issues here related to `toActorRef`. This wasn't tested by me at the moment but fixing this shouldn't be overly hard. We might want to ensure that this covers cases like this - at the moment we have a weird mix of `BaseActorRef` and `ActorRef` in the signature of this helper and I think there is a room for improvement there.

> The downside is that with the new typegen I've found it quite hard to use Generics in the context/events (form data structure in this example). We have a 300 max-line eslint rule at work so I often have to split a machine into multiple file which means that typeof machine becomes un-generic.

From my perspective lint rules are supposed to **help** the team. This is IMHO a clear place where this particular rule makes it way harder to implement stuff for you. So IMHO the real question here is - is this rule, in this context, a good rule? Should you continue using at all costs? Or perhaps this is a place where you should add an ignore comment for it? 

> I work around that by mangling the types at the hook border - outside the hook has good generic types, inside has static types. The disadvantage of this is that I haven't figured out how to pass the typegen out through the generics, though I'm tending to try avoid that situation and basically make the "Options" of the hook generic and properly typed and fudge some types on the output of the hook that don't need the generic inference.

I can see how passing around generics and types related to this is quite cumbersome but I don't fully understand how did you work around this issue for now.

I also think that if you need to pass this around then you should stick to using our helper types like `EventFrom`, `ActorRefFrom`, etc. If the existing ones are not sufficient for you then let's create a focused issue about your use case and discuss it there.

**Reply by SimeonC** [reply]
Thanks for your feedback. I understood the length description in the end, it was just my friday brain shutting down for the weekend that struggled.

The workaround for using the `fallbackActorRef` looks interesting - I might give it a shot if my current way doesn't pan out.

> From my perspective lint rules are supposed to help the team.

Fair point, I'm actually mostly in charge of maintaining our linter rules so I'll definitely look into disabling the eslint rule in someway and see if that helps. 😆 

> I can see how passing around generics and types related to this is quite cumbersome but I don't fully understand how did you work around this issue for now.

Basically what I did was write the machine as if it doesn't have generics (or rather hard code them), then all the hooks and return values of the hooks use the generics properly so type inference works when using the hooks, even though it's not entirely "correct" in it's implementation. Something like this;

```ts
const machine = createMachine({
  schema: {
    context: {} as Context<Something>;
  }
});

function useSomethingMachine<T extends Something>(option: T): Service<Context<T>, ...> {
  return useInterpret(machine.withContext({ option: option as Something }));
}
```

**Reply by SimeonC** [reply]
I opened a discussion about generic types here https://github.com/statelyai/xstate/discussions/3164. We can consider this discussion closed as the above suggestions should cover the use case I started with.
